### PR TITLE
issue #3905 Refactor to use ContactStore.getOneWithAllPubkeys() instead of .get() 

### LINKED ISSUE #3905: refactor to use ContactStore.getOneWithAllPubkeys instead of .get
as a part of #3897, discussed in https://github.com/FlowCrypt/flowcrypt-browser/pull/3898:

`Promise<Contact>` is actually used for subsequent rendering in `renderPubkeyResult`. If we want to move the "diagnostics" logic to compose-recipients module `evaluateRecipients` method, that's ok, but we don't have to present the results as a list of `Contact`, as this is a bit misleading (`email` field will be copied all over the entries), so the perfect solution would be a special type with `email` field and a list of keys.
Currently, the diagnostics (figuring out the properties of the most appropriate key) is done by the line below.
```
const [preferred] = await ContactStore.get(undefined, [email]);
```
We can, of cause, get rid of the compatibility method `ContactStore.get` and move the key-picking logic to some other place, but we need to find all places where it used.
I see it is used in `GmailElementReplacer.evaluateStandardComposeRecipients()` too

_Originally posted by @rrrooommmaaa in https://github.com/FlowCrypt/flowcrypt-browser/pull/3898#discussion_r686244628_

> so the perfect solution would be a special type with email field and a list of keys

That sounds good to me

> We can, of cause, get rid of the compatibility method ContactStore.get and move the key-picking logic to some other place, but we need to find all places where it used.
I see it is used in GmailElementReplacer.evaluateStandardComposeRecipients() too

Got it. Here's the code:

```ts
      // pick first usableForEncryption
      let selected = contactWithAllPubkeys.sortedPubkeys.find(entry => !entry.revoked && entry.pubkey.usableForEncryption);
      if (!selected) {
        selected = contactWithAllPubkeys.sortedPubkeys.find(entry => !entry.revoked && entry.pubkey.usableForEncryptionButExpired);
      }
      if (!selected) {
        selected = contactWithAllPubkeys.sortedPubkeys[0];
      }
```

`GmailElementReplacer.evaluateStandardComposeRecipients()` only needs to know if there is any usable key or not. So it could be as simple as:

```diff
-const [contact] = await ContactStore.get(undefined, [email]);
+const pubkeys = await ContactStore.getOneWithAllPubkeys(undefined, [email]);
-if (contact && contact.pubkey) {
+if (pubkeys.filter(p => p.usableForEncryption).length) {
```

Therefore, the actual nuanced logic that relates to choosing public keys for encryption can be actually in the compose modules somewhere. Then we can always use `.getOneWithAllPubkeys`, and never `.get`, which should produce fewer quirks in code and in behavior, I think.

-----------------

We'll probably want to rename `.get` to `.getByFingerprint` and then forbid using emails in the method, only fingerprints.

Most usages are by email. Some of these are just trying to understand if the recipient has encryption or not. Some actually need the key. We'd have to evaluate them individually.

image

Maybe another method would be useful, like `ContactStore.doesUsableKeyExistFor(email)` that could be used in `GmailElementReplacer.evaluateStandardComposeRecipients()` and also other places. Like this one:

```ts
      const [contact] = await ContactStore.get(undefined, [String($('.input_email').val())]);
      $('.action_add_contact')
        .text(contact?.hasPgp ? 'update key' : `import ${this.isExpired ? 'expired ' : ''}key`)
        .css('background-color', this.isExpired ? '#989898' : '');
    }
```

---------------------

> `Contact.hasPgp` just shows that a key exists, it may be revoked or expired, but generally I don't mind against moving key-selection logic out of ContactStore. It was implemented to mimick the original ContactStore with the intent of not making too much changes at once. Now that the code is looking stable, we can move on with further refactoring.

### PR DESCRIPTION
**Pull request by IvanPizhenko**
This PR refactors code to use `ContactStore.getOneWithAllPubkeys()` instead of `ContactStore.get()`

close #3905 

----------------------------------

**Tests** _(delete all except exactly one)_:
- Does not need tests (refactor only, docs or internal changes)

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

**Comment by IvanPizhenko** [comment]
@rrrooommmaaa Ready for review

File: extension/js/content_scripts/webmail/gmail-element-replacer.ts [hunk 1]
```diff
@@ -698,8 +698,8 @@ export class GmailElementReplacer implements WebmailElementReplacer {
               }
               if (typeof cache === 'undefined') {
                 try {
-                  const [contact] = await ContactStore.get(undefined, [email]);
-                  if (contact && contact.pubkey) {
+                  const contactWithPubKeys = await ContactStore.getOneWithAllPubkeys(undefined, email);
```
Comment (rrrooommmaaa) [inline extension/js/content_scripts/webmail/gmail-element-replacer.ts:701]: need to think this through. Should we try to lookup a pubkey if we only have a revoked version in the cache?...

**Review by rrrooommmaaa — CHANGES_REQUESTED** [review]

File: extension/chrome/elements/compose-modules/compose-recipients-module.ts [hunk 2]
```diff
@@ -461,8 +461,8 @@ export class ComposeRecipientsModule extends ViewModule<ComposeView> {
     this.addedPubkeyDbLookupInterval = Catch.setHandledInterval(async () => {
       const recipientsHasPgp: ValidRecipientElement[] = [];
       for (const recipient of noPgpRecipients) {
-        const [contact] = await ContactStore.get(undefined, [recipient.email]);
-        if (contact && contact.hasPgp) {
+        const contactWithPubKeys = await ContactStore.getOneWithAllPubkeys(undefined, recipient.email);
```
[the remark this answers is not in the document]
Comment (IvanPizhenko) [reply extension/chrome/elements/compose-modules/compose-recipients-module.ts:464]: Not sure how to check this, but will try

[the remark this answers is not in the document]
Comment (IvanPizhenko) [reply extension/chrome/elements/compose-modules/compose-recipients-module.ts:464]: Seems like `getEncryptionKeys()` does all the stuff.

**Comment by IvanPizhenko** [comment]
@rrrooommmaaa I've made some fixes. Please recheck.

File: test/source/tests/browser-unit-tests/unit-ContactStore.js [hunk 3]
```diff
@@ -318,9 +322,9 @@ BROWSER_UNIT_TEST_NAME(`ContactStore stores postfixed fingerprint internally for
   if (entity.fingerprint !== entityFp) {
     throw Error(`Failed to extract pubkey ${fingerprint}`);
   }
-  const [contactByEmail] = await ContactStore.get(db, [email]);
-  if (contactByEmail.pubkey.id !== fingerprint) {
-    throw Error(`Failed to extract pubkey ${fingerprint}`);
+  const contactByEmail = await ContactStore.getOneWithAllPubkeys(db, email);
+  if (contactByEmail.sortedPubkeys[0].pubkey.id !== fingerprint) {
```
Comment (rrrooommmaaa) [inline test/source/tests/browser-unit-tests/unit-ContactStore.js:326]: I have fixed this test

**Comment by IvanPizhenko** [comment]
@rrrooommmaaa all tests have passed, so PR likely can be merged

**Review by rrrooommmaaa — APPROVED** [review]

**Pull request merged**